### PR TITLE
batch write

### PR DESCRIPTION
**Pull request by kestfor**
## Summary

Batching significantly improves WAL write throughput (~2–6x, from ~210k to ~1.25M records/s) by amortizing per-operation overhead. Most of the gains are achieved with batch sizes of 100–1000; larger batches (e.g., 10k) provide only marginal improvements while greatly increasing memory usage and allocations. 
Per-record allocations remain roughly constant (~6–7 allocs, ~1.7–2.0KB), indicating that batching reduces syscall/overhead cost but does not optimize per-record memory behavior.

## Changes included

- Added new public struct:

```go
// Record represents a record in the log.
type Record struct {
    // Index is the unique index of the record for fast search.
    Index uint64
    // Key is the unique key of the record.
    Key   string
    // Value is the value of the record.
    Value []byte
}
```
WriteBatch now performs a single write syscall for the entire batch, significantly improving throughput.

## Impact
- Throughput increases ~2–6x compared to single-record writes.
- Optimal batch size: 100–1000 records.
- Memory and allocation overhead per record remain roughly unchanged.

### CODE REVIEW

File: wal.go [hunk 1]
```diff
@@ -167,44 +177,51 @@ func (c *Wal) CurrentIndex() uint64 {
...
+	defer c.mu.Unlock()
+	buf := bytes.NewBuffer(nil)
 
-	data, err := msgpack.Marshal(m)
-	if err != nil {
-		return errors.Wrap(err, "failed to encode msg")
+	for _, record := range batch {
+		if _, exists := c.index[record.Index]; exists {
+			return ErrExists
+		}
```
[the remark this answers is not in the document]
Comment (kestfor) [reply wal.go:206]: fixed

File: wal.go [hunk 2]
```diff
@@ -167,44 +177,51 @@ func (c *Wal) CurrentIndex() uint64 {
...
+			return err
+		}
+		m := msg{Key: record.Key, Value: record.Value, Idx: record.Index}
+		m.Checksum = m.calculateChecksum()
+		data, err := msgpack.Marshal(m)
+		if err != nil {
+			return errors.Wrap(err, "failed to encode msg")
+		}
+		buf.Write(data)
+		c.tmpIndex[record.Index] = m
```
[the remark this answers is not in the document]
Comment (kestfor) [reply wal.go]: changed

File: wal.go [hunk 3]
```diff
@@ -167,44 +177,51 @@ func (c *Wal) CurrentIndex() uint64 {
...
-	if err != nil {
-		return errors.Wrap(err, "failed to encode msg")
+	for _, record := range batch {
+		if _, exists := c.index[record.Index]; exists {
+			return ErrExists
+		}
+		if err := c.rotateIfNeeded(); err != nil {
+			return err
+		}
+		m := msg{Key: record.Key, Value: record.Value, Idx: record.Index}
```
[the remark this answers is not in the document]
Comment (kestfor) [reply wal.go:221]: changed

File: wal.go [hunk 4]
```diff
@@ -167,44 +177,85 @@ func (c *Wal) CurrentIndex() uint64 {
...
+			return ErrExists
+		}
 	}
 
-	if err := c.rotateIfNeeded(); err != nil {
-		return err
-	}
+	var buf bytes.Buffer
+	nextRotateAfter := c.segmentsThreshold - len(c.tmpIndex)
+	messages := make([]msg, 0, nextRotateAfter)
```
[the remark this answers is not in the document]
Comment (kestfor) [reply wal.go]: new cover tests were added and also checks for duplicates indexes inside batch and from c.index and c.tmpIndex

File: wal_test.go [hunk 5]
```diff
@@ -418,3 +421,160 @@ func TestWriteTombstone(t *testing.T) {
...
+func BenchmarkWal_WriteBatch100(b *testing.B)   { benchmarkWALWrite(b, 100) }
+func BenchmarkWal_WriteBatch1000(b *testing.B)  { benchmarkWALWrite(b, 1000) }
+func BenchmarkWal_WriteBatch10000(b *testing.B) { benchmarkWALWrite(b, 10000) }
+
+// Batching significantly improves WAL write throughput (~2–6x, from ~210k to ~1.25M records/s)
+// by amortizing per-operation overhead. Most of the gains are achieved with batch sizes of 100–1000;
+// larger batches (e.g., 10k) provide only marginal improvements while greatly increasing memory usage and allocations.
+//
+// Per-record allocations remain roughly constant (~6–7 allocs, ~1.7–2.0KB),
+// indicating that batching reduces syscall/overhead cost but does not optimize per-record memory behavior.
```
Comment (vadiminshakov) [inline wal_test.go]: No words needed, just create a benchmark with two b.Run cases.

File: wal_test.go [hunk 6]
```diff
@@ -418,3 +421,160 @@ func TestWriteTombstone(t *testing.T) {
...
+	for i := 0; i < len(records); i += batchSize {
+		end := min(i+batchSize, len(records))
+
+		if batchSize == 1 {
+			require.NoError(b, log.Write(records[i].Index, records[i].Key, records[i].Value))
+			total++
+		} else {
+			require.NoError(b, log.WriteBatch(records[i:end]))
+			total += end - i
+		}
```
Comment (vadiminshakov) [inline wal_test.go]: One benchmark must test exactly one execution path

File: wal_test.go [hunk 7]
```diff
@@ -418,3 +421,160 @@ func TestWriteTombstone(t *testing.T) {
...
+		require.Equal(t, r.Key, key)
+		require.Equal(t, r.Value, value)
+	}
+
+	// CurrentIndex should reflect the highest index
+	require.Equal(t, uint64(len(batch)), wal.CurrentIndex())
+}
+
+// getRandomData returns a random byte slice of the given size.
+func getRandomData(sizeBytes int) []byte {
```
Comment (vadiminshakov) [inline wal_test.go]: can we move bench to _bench file?

File: wal.go [hunk 8]
```diff
@@ -167,6 +182,92 @@ func (c *Wal) CurrentIndex() uint64 {
...
+			return ErrExists
+		}
+	}
+
+	c.buf.Reset()
+	messages := make([]msg, 0, c.segmentsThreshold)
+	c.buf.Grow(len(batch) * 128) // small heuristic; avoids repeated growth on small/medium batches
+	nextRotateAfter := c.segmentsThreshold - len(c.tmpIndex)
+
+	flush := func() error {
```
Comment (vadiminshakov) [inline wal.go]: can we move it to separate method and reuse in Write?

**Comment by vadiminshakov** [comment]
Great contribution, just needs a few minor fixes.

**Comment by kestfor** [comment]
Addressed remaining review point about CurrentIndex semantics in batch writes.

Changes pushed in commit 617ad73:
- WriteBatch now updates lastIndex by max written record index instead of +len(batch)
- Keeps behavior consistent with startup restore logic in NewWAL

Validation: go test ./... passes.

File: wal.go [hunk 9]
```diff
@@ -167,44 +177,51 @@ func (c *Wal) CurrentIndex() uint64 {
...
-	if _, exists := c.index[index]; exists {
-		return ErrExists
-	}
-
-	if err := c.rotateIfNeeded(); err != nil {
-		return err
+func (c *Wal) WriteBatch(batch []Record) error {
+	if len(batch) == 0 {
+		return nil
 	}
```
[the remark this answers is not in the document]
Comment (kestfor) [reply wal.go:189]: added

**Comment by kestfor** [comment]
Additional fix for WriteBatch rotation edge case.

Issue:
- When len(tmpIndex) == SegmentThreshold at WriteBatch entry, nextRotateAfter starts at 0 and then becomes negative in the loop, so rotation branch is never reached.

Fix (commit 4563a1a):
- Pre-rotate at WriteBatch start when nextRotateAfter <= 0.
- Reset nextRotateAfter to full segment threshold after pre-rotation.

Tests:
- Added TestWriteBatch_RotateWhenTmpIndexAlreadyFull to cover this scenario.
- go test ./... passes.

**Comment by vadiminshakov** [comment]
> Additional fix for WriteBatch rotation edge case.
> 
> Issue:
> 
> * When len(tmpIndex) == SegmentThreshold at WriteBatch entry, nextRotateAfter starts at 0 and then becomes negative in the loop, so rotation branch is never reached.
> 
> Fix (commit [4563a1a](https://github.com/vadiminshakov/gowal/commit/4563a1afb5d6f93daf3ac463e1c017a5d004018d)):
> 
> * Pre-rotate at WriteBatch start when nextRotateAfter <= 0.
> * Reset nextRotateAfter to full segment threshold after pre-rotation.
> 
> Tests:
> 
> * Added TestWriteBatch_RotateWhenTmpIndexAlreadyFull to cover this scenario.
> * go test ./... passes.

How can we end up in a state where len(tmpIndex) == SegmentThreshold? Write rotates the segment when the threshold is reached, and WriteBatch used to do that too.

**Comment by kestfor** [comment]
> > Additional fix for WriteBatch rotation edge case.
> > Issue:
> > 
> > * When len(tmpIndex) == SegmentThreshold at WriteBatch entry, nextRotateAfter starts at 0 and then becomes negative in the loop, so rotation branch is never reached.
> > 
> > Fix (commit [4563a1a](https://github.com/vadiminshakov/gowal/commit/4563a1afb5d6f93daf3ac463e1c017a5d004018d)):
> > 
> > * Pre-rotate at WriteBatch start when nextRotateAfter <= 0.
> > * Reset nextRotateAfter to full segment threshold after pre-rotation.
> > 
> > Tests:
> > 
> > * Added TestWriteBatch_RotateWhenTmpIndexAlreadyFull to cover this scenario.
> > * go test ./... passes.
> 
> How can we end up in a state where len(tmpIndex) == SegmentThreshold? Write rotates the segment when the threshold is reached, and WriteBatch used to do that too.

In the write function, rotation occurs before the write itself, for example, let's say SegmentThreshold = 3:
1. Write(1)
2. Write(2)
3. Write(3)
After this, len(tmpIndex) == 3 and there is no rotation yet, because rotateIfNeeded() is called before the write, not after.
4. Then call WriteBatch(...) and enter with len(tmpIndex) == SegmentThreshold.

Therefore, the batch operation must check this case and handle it correctly.

**Comment by vadiminshakov** [comment]
> 3\. Write(3)
>     After this, len(tmpIndex) == 3 and there is no rotation yet, because rotateIfNeeded() is called before the write, not after.

No, tmpIndex is recreated after the segment rotation.

https://github.com/vadiminshakov/gowal/blob/main/segment.go#L51

**Comment by kestfor** [comment]
> > 3. Write(3)
> > After this, len(tmpIndex) == 3 and there is no rotation yet, because rotateIfNeeded() is called before the write, not after.
> 
> No, tmpIndex is recreated after the segment rotation.
> 
> https://github.com/vadiminshakov/gowal/blob/main/segment.go#L51

didn't get your point, so I wrote a new test that proves my example, you can check it out: https://github.com/vadiminshakov/gowal/pull/2/changes#diff-1b8abbc450c378dbd952de9a85574d8501b0c9aded5ac60ed469dee331ee045dR543

**Comment by vadiminshakov** [comment]
> > > 3. Write(3)
> > >    After this, len(tmpIndex) == 3 and there is no rotation yet, because rotateIfNeeded() is called before the write, not after.
> > 
> > 
> > No, tmpIndex is recreated after the segment rotation.
> > https://github.com/vadiminshakov/gowal/blob/main/segment.go#L51
> 
> didn't get your point, so I wrote a new test that proves my example, you can check it out: https://github.com/vadiminshakov/gowal/pull/2/changes#diff-1b8abbc450c378dbd952de9a85574d8501b0c9aded5ac60ed469dee331ee045dR543

Got it, then the check is needed there — you're right.

File: wal.go [hunk 10]
```diff
@@ -167,6 +182,102 @@ func (c *Wal) CurrentIndex() uint64 {
...
+	}
+
+	flush := func() error {
+		if c.buf.Len() == 0 {
+			return nil
+		}
+
+		if _, err := c.log.Write(c.buf.Bytes()); err != nil {
+			return errors.Wrap(err, "failed to write msg to log")
+		}
```
Comment (vadiminshakov) [inline wal.go]: If the write here is only partially successful for some of the records in the batch, the operation will end in an inconsistent state: the tmpIndex will not receive even the records that were partially written from the batch, and lastIndex will remain in the same state as before the batch write.

File: wal.go [hunk 11]
```diff
@@ -167,6 +182,102 @@ func (c *Wal) CurrentIndex() uint64 {
 	return c.lastIndex.Load()
 }
 
+// WriteBatch appends a batch of records to the WAL in a single operation.
+func (c *Wal) WriteBatch(batch []Record) error {
```
Comment (vadiminshakov) [inline wal.go:186]: After a partial write, the caller won't know what to do next, because it won't know which records were written and which were not.

**Comment by vadiminshakov** [comment]
We can make all of this much simpler: don’t try to split the batch, and write it atomically instead. The code will be much simpler and far more reliable.

**Comment by kestfor** [comment]
> We can make all of this much simpler: don’t try to split the batch, and write it atomically instead. The code will be much simpler and far more reliable.

So you're suggesting that SegmentThreshold should become a soft limit? otherwise I don't see any way to make batch writes an atomic operation.

**Comment by vadiminshakov** [comment]
> > We can make all of this much simpler: don’t try to split the batch, and write it atomically instead. The code will be much simpler and far more reliable.
> 
> So you're suggesting that SegmentThreshold should become a soft limit? otherwise I don't see any way to make batch writes an atomic operation.

Yes, it's ok for BatchWrite

File: wal_test.go (hunk 5 above)
Comment (kestfor) [reply wal_test.go]: rdy

File: wal_test.go (hunk 6 above)
Comment (kestfor) [reply wal_test.go]: divided into separated functioins

File: wal_test.go (hunk 7 above)
Comment (kestfor) [reply wal_test.go]: rdy

File: wal.go (hunk 8 above)
Comment (kestfor) [reply wal.go]: made write method to call writebanch since there is not additional rotation logic, also that lightly improved allocs

File: wal.go (hunk 10 above)
Comment (kestfor) [reply wal.go]: fixed

File: wal.go (hunk 11 above)
Comment (kestfor) [reply wal.go:186]: Now it's all or nothing.

**Comment by kestfor** [comment]
kind of ready i think, took all the comments into account

File: wal_test.go [hunk 12]
```diff
@@ -418,3 +419,121 @@ func TestWriteTombstone(t *testing.T) {
...
+	batch := []Record{
+		{Index: 5, Key: "key5-conflict", Value: []byte("valueB")},
+		{Index: 6, Key: "key6", Value: []byte("value6")},
+	}
+
+	err = wal.WriteBatch(batch)
+	require.ErrorIs(t, err, ErrExists, "expected ErrExists for index conflict with existing WAL entry")
+}
+
+func TestWriteBatch_AcrossSegmentThreshold(t *testing.T) {
```
Comment (vadiminshakov) [inline wal_test.go:467]: This test is no longer needed.

File: wal_test.go [hunk 13]
```diff
@@ -418,3 +419,121 @@ func TestWriteTombstone(t *testing.T) {
...
+		require.NoError(t, err, "expected record to exist after batch write and rotation")
+		require.Equal(t, r.Key, key)
+		require.Equal(t, r.Value, value)
+	}
+
+	// CurrentIndex should reflect the highest index
+	require.Equal(t, uint64(len(batch)), wal.CurrentIndex())
+}
+
+func TestWriteBatch_RotateWhenTmpIndexAlreadyFull(t *testing.T) {
```
Comment (vadiminshakov) [inline wal_test.go:505]: No need to test rotation here
